### PR TITLE
Support quantized tensor subclasses across Python and C++

### PR DESCRIPTION
**Pull request by dingqingy-nv**
# Description

Transformer Engine quantized tensors can be subclassed, but two boundaries assumed an exact concrete wrapper class:

1. Concrete `detach()` implementations reconstructed the base TE class, dropping the tensor's runtime subclass. PyTorch then rejected the result when rewrapping it as a `torch.nn.Parameter`, because `Parameter` requires `detach()` to preserve the exact runtime type.
2. The PyTorch C++ bindings used exact `Py_TYPE` comparisons for quantized tensor outputs. As a result, operations such as `MXFP8Quantizer.update_quantized(src, dst)` rejected a valid `MXFP8Tensor` subclass even though Python `isinstance(dst, MXFP8Tensor)` was true.

Megatron Core GTP exposes both cases because it represents native quantized parameters with dynamic subclasses such as `GTP_MXFP8Tensor`. The newer module application path in #3153 made the latent detach mismatch visible. The downstream integration is NVIDIA/Megatron-LM#6546.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Define runtime-type preservation as part of the `QuantizedTensor.detach()` contract.
- Construct detached tensors through the runtime class for `Float8Tensor`, `Float8BlockwiseQTensor`, `MXFP8Tensor`, `NVFP4Tensor`, `IdentityTensor`, and `HybridQuantizedTensor`.
- Use CPython subtype checks for the Float8, blockwise Float8, MXFP8, and NVFP4 tensor/storage wrapper families in the C++ bindings. Quantizer objects retain their existing exact-type checks.
- Add tests that verify exact subclass preservation, aliased quantized storage, detached autograd state, successful `torch.nn.Parameter` construction, and `update_quantized` into dynamic subclasses.

## Validation

Built and tested on GB300 from TE commit `bf64b4e8b2985ce7ff394b7f3cb240e764b24a3a`:

- 10/10 quantized subclass regression cases passed, covering detach and C++ `update_quantized` across Float8, blockwise Float8, MXFP8, NVFP4, Identity, and Hybrid wrappers.
- 3/3 existing module `_apply` attribute-preservation tests passed.
- 135/135 MXFP8 2D tests passed.
- Transformer Engine Python and C++ lint passed.
- Clean Megatron Core focused suite passed 10/10 with the downstream detach/update workarounds absent.
- Two-node real-data 1D and 2D MXFP8 proxy runs each completed 10 training iterations plus validation with GTP and native MXFP8 parameter gather.
- The same two-node 1D and 2D runs also passed with `NVTE_CUTEDSL_FUSED_GROUPED_MLP=1` after installing the official final cuDNN-frontend 1.27.0 package:
  - [1D W&B run](https://wandb.ai/nvidia/dingqingy-8-17/runs/yiz9iak8)
  - [2D W&B run](https://wandb.ai/nvidia/dingqingy-8-17/runs/b8ix9ar0)

The cuDNN-frontend package replacement addresses a separate pre-release package mismatch around optional `prob_tensor`; it is not part of this PR.

# Checklist

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/tensor/float8_tensor.py [hunk 1]
```diff
@@ -521,7 +521,7 @@ def quantize_(
 
     def detach(self) -> Float8Tensor:
         # pylint: disable=missing-function-docstring
-        return Float8Tensor.make_like(self)
+        return self.__class__.make_like(self)
```
Comment (ptrendx) [inline transformer_engine/pytorch/tensor/float8_tensor.py]: Could you create this as the default implementation in QuantizedTensor instead?

File: transformer_engine/pytorch/csrc/pybind.h [hunk 2]
```diff
@@ -57,13 +57,15 @@ inline bool IsFloat8CurrentScalingQuantizers(PyObject *obj) {
 }
 
 inline bool IsFloat8Tensor(PyObject *obj) {
-  return Py_TYPE(obj) == Float8TensorPythonClass || Py_TYPE(obj) == Float8TensorStoragePythonClass;
+  return PyObject_TypeCheck(obj, Float8TensorPythonClass) ||
```
Comment (ptrendx) [inline transformer_engine/pytorch/csrc/pybind.h:60]: I don't like this as this introduces addition CPU overhead on every call into the C extensions. Why do you need this?

File: transformer_engine/pytorch/quantized_tensor.py [hunk 3]
```diff
@@ -798,7 +798,7 @@ def detach(self) -> QuantizedTensor:
         """Create new quantized tensor with same data
 
         Output tensor must be detached from the current autograd
-        graph.
+        graph and have the same runtime type as ``self``.
```
Comment (ptrendx) [inline transformer_engine/pytorch/quantized_tensor.py]: This additional remark is not really needed.

File: transformer_engine/pytorch/csrc/pybind.h (hunk 2 above)
Comment (dingqingy-nv) [reply transformer_engine/pytorch/csrc/pybind.h:60]: The PyTorch C++ bindings used exact `Py_TYPE `comparisons for quantized tensor outputs. As a result, operations such as `MXFP8Quantizer.update_quantized(src, dst)` rejected a valid MXFP8Tensor subclass even though Python `isinstance(dst, MXFP8Tensor)` was true.

File: transformer_engine/pytorch/tensor/float8_tensor.py (hunk 1 above)
Comment (dingqingy-nv) [reply transformer_engine/pytorch/tensor/float8_tensor.py]: Yes, will do

File: transformer_engine/pytorch/quantized_tensor.py (hunk 3 above)
Comment (dingqingy-nv) [reply transformer_engine/pytorch/quantized_tensor.py]: will remove

File: transformer_engine/pytorch/csrc/pybind.h (hunk 2 above)
Comment (dingqingy-nv) [reply transformer_engine/pytorch/csrc/pybind.h:60]: will preserve the current exact `Py_TYPE` path for ordinary TE tensors, and only run subtype-aware checks after exact dispatch fails

Comment (ptrendx) [reply transformer_engine/pytorch/csrc/pybind.h:60]: To be honest, I don't think that would change anything. Let's keep it with the TypeCheck then and we will monitor the additional overhead.

Comment (dingqingy-nv) [reply transformer_engine/pytorch/csrc/pybind.h:60]: sounds good

**Review by ptrendx — APPROVED** [review]
